whylog/teacher/constraint_links_base.py

```diff
@@ -0,0 +1,48 @@
+class ConstraintLinksBase:
+    """
+    Container for links between group and constraint
+    """
+
+    def __init__(self, links_list=None):
+        """
+        :param links_list: list of links between group and constraint
+                          one link is a tuple (line_id, group_no, constraint_id)
+        :type links_list: list[(int, int, int)]
+        """
+        self.links = links_list or []
+        self.constr_id_position = 2
+
+    def get_links(self):
+        return self.links
+
+    def distinct_constraint_ids(self):
+        return set(list(zip(*self.links))[self.constr_id_position])
```

Comment:
`:type links_list: list[(int, int, int)]` so `zip(*self.links)` is (in python3 iterator of) 3 lists of ints.
`list(zip(*self.links)` - converts iterator into list of 3 lists
`list(zip(*self.links))[self.constr_id_position]` - we take third list - containing constraints ids
`set(list(zip(*self.links))[self.constr_id_position])` - we take distinct constraint ids